pkg/api/store/nodetemplate/store.go

```diff
@@ -24,43 +24,95 @@ type nodeTemplateStore struct {
 	types.Store
 	NodePoolLister        v3.NodePoolLister
 	CloudCredentialLister corev1.SecretLister
-	NodeTemplateClient    v3.NodeTemplateInterface
+}
+
+type transformer struct {
+	NodeTemplateClient v3.NodeTemplateInterface
 }
 
 func Wrap(store types.Store, npLister v3.NodePoolLister, secretLister corev1.SecretLister, ntClient v3.NodeTemplateInterface) types.Store {
-	s := &nodeTemplateStore{
-		Store:                 store,
+	t := &transformer{
+		NodeTemplateClient: ntClient,
+	}
+	s := &transform.Store{
+		Store:       store,
+		Transformer: t.filterLegacyTemplates,
+	}
+	return &nodeTemplateStore{
+		Store:                 s,
 		NodePoolLister:        npLister,
 		CloudCredentialLister: secretLister,
-		NodeTemplateClient:    ntClient,
-	}
-	return &transform.Store{
-		Store:       s,
-		Transformer: s.filterLegacyTemplates,
 	}
 }
 
-func (s *nodeTemplateStore) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
+func (t *transformer) filterLegacyTemplates(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, opt *types.QueryOptions) (map[string]interface{}, error) {
 	ns, _ := data["namespaceId"].(string)
-	name, _ := data["name"].(string)
+	id, _ := data["id"].(string)
 	if ns != namespace.NodeTemplateGlobalNamespace {
-		s.NodeTemplateClient.Controller().Enqueue(ns, name)
+		// nodetemplate may not have been migrated yet, enqueue
+		t.NodeTemplateClient.Controller().Enqueue(ns, strings.TrimPrefix(id, ns+":"))
 		return nil, nil
 	}
 	return data, nil
 }
 
+func (s *nodeTemplateStore) ByID(apiContext *types.APIContext, schema *types.Schema, id string) (map[string]interface{}, error) {
+	originalNs, originalID := getNsAndID(id)
+	fullMigratedID := getMigratedID(originalNs, originalID)
+	migratedNs, migratedID := getNsAndID(fullMigratedID)
+	data, err := s.Store.ByID(apiContext, schema, fullMigratedID)
+	if err != nil {
+		return nil, replaceIDInError(err, migratedID, originalID, migratedNs, originalNs)
```

Comment:
if s.Store.ByID returns say k8serror IsNotFound, their can be details about the legacy ID in the error, do we handle them too in replaceIDInError?

Reply:
I believe in the proxy store k8s errors (which fit the interface Status) are translated to norman api errors: https://github.com/rancher/norman/blob/07a7713a06d92f6a2aee25cb2b86c6670577febc/store/proxy/error_wrapper.go#L46